src/mpfao/3D/tdympfao3D_ts.c

```diff
@@ -373,6 +389,200 @@ PetscErrorCode TDyMPFAOIJacobian_Vertices_3DMesh(Vec Ul, Mat A, void *ctx) {
+    //                     rho_ij       + d((kr/mu)_{ij,upwind})/dP_k * [T] * [P] +
+    //                     rho_ij       +   (kr/mu)_{ij,upwind}       * T_k       +
+    //                  2 * rho_ij * d(rho_ij)/dP_k *   (kr/mu)_{ij,upwind}       * G  +
+    //                      rho_ij *   rho_ij       * d((kr/mu)_{ij,upwind})/dP_k * G
+    //
+    // For k /= i or j, the jacobian is given as:
+    // d(fluxm_ij)/dP_k =
+    //                     rho_ij       +   (kr/mu)_{ij,upwind}       * T_k
+    //
+    for (PetscInt irow=0; irow<nflux_in + npitf_bc; irow++) {
+
+      PetscInt face_id = vertices->face_ids[vOffsetFace + irow];
+      PetscInt subface_id = vertices->subface_ids[vOffsetFace + irow];
+      PetscInt fOffsetCell = faces->cell_offset[face_id];
+
+      PetscInt cell_id_up = faces->cell_ids[fOffsetCell + 0];
+      PetscInt cell_id_dn = faces->cell_ids[fOffsetCell + 1];
+
+      // If using neumann bc (which is currently no-flux), then skip the face
+      if ( tdy->mpfao_bc_type == MPFAO_NEUMANN_BC  && (cell_id_up<0 || cell_id_dn <0))  continue;
+
+      PetscReal dukvr_dPup = 0.0;
+      PetscReal dukvr_dPdn = 0.0;
+      PetscReal ukvr;
+
+      PetscReal den = 0.0;
+      if (cell_id_up>=0) den += tdy->rho[cell_id_up];
+      else               den += tdy->rho_BND[-cell_id_up-1];
+      if (cell_id_dn>=0) den += tdy->rho[cell_id_dn];
+      else               den += tdy->rho_BND[-cell_id_dn-1];
+      den *= 0.5;
+
+      PetscInt num_subfaces = 4;
+      if (TtimesP[irow] + den*GravDis_ptr[face_id*num_subfaces + subface_id]< 0.0) { // up ---> dn
+        // Flow: up --> dn
+        if (cell_id_up>=0) {
+          // "up" is an internal cell
+          ukvr       = cc->Kr[cell_id_up]/tdy->vis[cell_id_up];
+          dukvr_dPup = cc->dKr_dS[cell_id_up]*cc->dS_dP[cell_id_up]/tdy->vis[cell_id_up] -
+                       cc->Kr[cell_id_up]/(tdy->vis[cell_id_up]*tdy->vis[cell_id_up])*tdy->dvis_dP[cell_id_up];
+        } else {
+          // "up" is boundary cell
+          ukvr = cc_bnd->Kr[-cell_id_up-1]/tdy->vis_BND[-cell_id_up-1];
+          dukvr_dPup = 0.0;//cc->dKr_dS[-cell_id_up-1]*cc_bnd->dS_dP_BND[-cell_id_up-1]/tdy->vis_BND[-cell_id_up-1];
+        }
+
+      } else {
+        // Flow: up <--- dn
+        if (cell_id_dn>=0) {
+          // "dn" is an internal cell
+          ukvr       = cc->Kr[cell_id_dn]/tdy->vis[cell_id_dn];
+          dukvr_dPdn = cc->dKr_dS[cell_id_dn]*cc->dS_dP[cell_id_dn]/tdy->vis[cell_id_dn] -
+                       cc->Kr[cell_id_dn]/(tdy->vis[cell_id_dn]*tdy->vis[cell_id_dn])*tdy->dvis_dP[cell_id_dn];
+        } else {
+          // "dn" is a boundary cell
+          ukvr       = cc_bnd->Kr[-cell_id_dn-1]/tdy->vis_BND[-cell_id_dn-1];
+          dukvr_dPdn = 0.0;//cc_bnd->dKr[-cell_id_dn-1]*cc_bnd->dS_dP[-cell_id_dn-1]/tdy->vis_BND[-cell_id_dn-1];
+        }
+      }
+
+      // If one of the cell is on the boundary
+      if (cell_id_up<0) cell_id_up = cell_id_dn;
```

Comment:
If you don't mind, let's use curly braces for `if` tests, even if they're one-liners. This makes it easier for people to add logic.